File: sp-2 security/SCM/src/main/java/com/example/SCM/service/ProductRequirementService.java
package com.example.SCM.service;

import com.example.SCM.dto.request.ProductRequirementRequestDTO;
import com.example.SCM.dto.response.ProductRequirementResponseDTO;
import java.util.List;
import java.util.Optional;

public interface ProductRequirementService {

    ProductRequirementResponseDTO save(ProductRequirementRequestDTO dto);

    ProductRequirementResponseDTO update(Long id, ProductRequirementRequestDTO dto);

    ProductRequirementResponseDTO updateStatus(Long id, String status);

    List<ProductRequirementResponseDTO> findAll();

    Optional<ProductRequirementResponseDTO> getById(Long id);

    void delete(Long id);
}

File: Anguler/SCM_Anguler/src/app/service/product-requirement.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';
import { environment } from '../../environment/environment';
import { ProductRequirementRequest, ProductRequirementResponse } from '../component/shared/model/productRequirementModel';

@Injectable({
  providedIn: 'root'
})
export class ProductRequirementService {

  private apiUrl = environment.apiUrl + 'product-requirements';

  constructor(private http: HttpClient) { }

  findAll(): Observable<ProductRequirementResponse[]> {
    return this.http.get<ProductRequirementResponse[]>(this.apiUrl);
  }

  getById(id: number): Observable<ProductRequirementResponse> {
    return this.http.get<ProductRequirementResponse>(`${this.apiUrl}/${id}`);
  }

  save(dto: ProductRequirementRequest): Observable<ProductRequirementResponse> {
    return this.http.post<ProductRequirementResponse>(this.apiUrl, dto);
  }

  update(id: number, dto: ProductRequirementRequest): Observable<ProductRequirementResponse> {
    return this.http.put<ProductRequirementResponse>(`${this.apiUrl}/${id}`, dto);
  }

  updateStatus(id: number, status: string): Observable<ProductRequirementResponse> {
    return this.http.patch<ProductRequirementResponse>(`${this.apiUrl}/${id}/status`, null, {
      params: { status }
    });
  }

  delete(id: number): Observable<string> {
    return this.http.delete(`${this.apiUrl}/${id}`, { responseType: 'text' });
  }
}
